tests/bot/rules/test_mentions.py

```diff
@@ -0,0 +1,99 @@
+import unittest
+from typing import List, NamedTuple, Tuple
+
+from bot.rules import mentions
+from tests.helpers import async_test
+
+
+class FakeMessage(NamedTuple):
+    author: str
+    mentions: List[None]
+
```

Comment:
Why are you not using the `MockMessage` defined in [`tests.helpers`](../../blob/master/tests/helpers.py#L332)? If it's not important to have a `MockMember` for author attribute, you can just provide a string for the author field like you do now: `MockMessage(author="bob", mentions=list(range(3)))` would work and create a mock that follows the specifications of an actual `discord.Message` object.

Reply:
I'm not sure if that's desirable though, because in that case the `__eq__` will [look at the Message's id](https://github.com/Rapptz/discord.py/blob/4c3e53edf4ff90e3fcf195c429601a553d661cd8/discord/mixins.py#L31). So we would need to be manually controlling the `id` attribute to correctly compare the output tuples. Using a `FakeMessage`, we simply compare them based on the attribs we need for the test. I believe using `MockMessage` would introduce needless complexity as a result.

This idea is further explained in [the unit test for the attachments rule](https://github.com/python-discord/bot/blob/3953aad9c21e5593a980ed996f54dccdbfac36dd/tests/bot/rules/test_attachments.py#L9) which I draw a lot of inspiration from.

Reply:
[`discord.Message`](https://github.com/Rapptz/discord.py/blob/4c3e53edf4ff90e3fcf195c429601a553d661cd8/discord/message.py#L176) does not inherit from [`EqualityComparable`](https://github.com/Rapptz/discord.py/blob/4c3e53edf4ff90e3fcf195c429601a553d661cd8/discord/mixins.py#L27).

Reply:
I'm a bit torn on the complexity argument here.

One of the reasons we use our custom mock objects is because they follow the specifications of an actual instance of the `discord.py` they are mocking. This means that if a version upgrade in `discord.py` changes the API of the class/object in a way that's relevant to the code you're testing, the tests will now fail because the code it's testing is trying to use attributes that no longer exist on the actual object.

By creating a static object that is a snapshot of the current API, you bypass that spec safeguard of our mock objects: Tests will not show that the code is using no-longer-existing attributes, because your `FakeMessage` happily provides an object with the old attributes for the test.

Another point is that we're now back to creating multiple very similar classes to mock the same object in a lot of different files while we have a single helper mock that could be used for all of them. I'm not sure what I find more complex in the end: Having a lot of slightly different classes in a lot of files or a single mock type that works like the actual object we're trying to mock.

Granted, the current mock we have does not implement an `__eq__` method, which makes it so that our mocks are not compared in a similar way to actual message objects. I think a good way forward is implementing a `__eq__` similar to the `__eq__` the `discord.Message` inherits and modifying the `__init__` to make sure we set an `id` at initialization. We could even an auto-incremental ID to make sure each new mock gets a new message id automatically, which will mean we don't have to worry about controlling the `id` in individual tests, unless we want it to be an exact value.

---
Just as a clarification of how the mocks work:

 > because in that case the __eq__ will look at the Message's id.

This not actually the case. The mocks follow the specifications of the `discord.py`, which means that trying to access attributes that do not exist on the original will result in an `AttributeError`. They do not delegate the methods to the original object or subclass the original object, so using an equality operator will not result in the `__eq__` method of the original `discord.py` class being called. That means we'd have to indeed control the `__eq__` and that's not really necessary here.

Reply:
@scragly Yes, sorry about that, I should have checked before assuming. It looks like `discord.Message` does not have a custom `__eq__` implementation, but I suppose the point is that the `id` obviously plays a role.

@SebastiaanZ Many thanks for the feedback, I will adjust the test cases as per our discussion on discord.

Reply:
In https://github.com/python-discord/bot/pull/655/commits/16e6c36c6b370300ef7507d2e01421ad0d83f407, `tests.helpers.MockMessage` is used as suggested. Passing an `id` (message id, that is) manually actually ends up being unnecessary, as the `__eq__` falls back on identity comparison. The test cases just needed to be adjusted so that multiple instances aren't used to represent one message.

Let me know what you think. If this looks good, I will adjust the other rule unit tests to use this methodology.

Reply:
@SebastiaanZ I'm marking this as resolved since we're now using `MockMessage`, feel free to un-resolve if there's more you'd like me to change.